Data/ByteString/Builder/RealFloat.hsc

```diff
@@ -0,0 +1,302 @@
+
+foreign import ccall unsafe "static f2s_to_chars"
+    c_ryu_f2s_to_chars :: CUInt -> CInt -> CUChar -> Ptr Word8 -> IO CInt
+
+foreign import ccall unsafe "static d2s_to_chars"
+    c_ryu_d2s_to_chars :: CULong -> CInt -> CUChar -> Ptr Word8 -> IO CInt
+
+#include "ryu.h"
+
+{-# INLINABLE f2s_max_digits #-}
+f2s_max_digits :: Int
+f2s_max_digits = #const F2S_MAX_DIGITS
+
+{-# INLINABLE d2s_max_digits #-}
+d2s_max_digits :: Int
+d2s_max_digits = #const D2S_MAX_DIGITS
+
+{-# INLINE ryu_f2s #-}
+ryu_f2s :: Float -> ByteString
+ryu_f2s f = unsafeDupablePerformIO $ do
+    fp <- mallocByteString f2s_max_digits :: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_f2s (realToFrac f) p
+
+{-# INLINE ryu_d2s #-}
+ryu_d2s :: Double -> ByteString
+ryu_d2s f = unsafeDupablePerformIO $ do
+    fp <- mallocByteString d2s_max_digits :: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_d2s (realToFrac f) p
+
+data FloatingDecimal64 = FD64 !Word64 !Int32
+data FloatingDecimal32 = FD32 !Word32 !Int32
+
+instance Show FloatingDecimal64 where
+  showsPrec p (FD64 m e) = showsPrec p m `mappend` showsPrec p '.' `mappend` showsPrec p e
+
+instance Show FloatingDecimal32 where
+  showsPrec p (FD32 m e) = showsPrec p m `mappend` showsPrec p '.' `mappend` showsPrec p e
+
+-- extracts base-10 converted mantissa and exponent for floats using ryu
+-- algorithm
+-- NB: only valid if not NaN, +/-0, or +/-Inf. In practice, all calls should
+--     guarded by `specialStr`
+{-# INLINE ryu_f2s_fd #-}
+ryu_f2s_fd :: Float -> FloatingDecimal32
+ryu_f2s_fd f = unsafeDupablePerformIO $
+    alloca $ \mOut -> do
+      alloca $ \eOut -> do
+        c_ryu_f2s_fd (realToFrac f) mOut eOut
+        m <- peek mOut
+        e <- peek eOut
+        return $ FD32 m e
+
+-- extracts base-10 converted mantissa and exponent for doubles using ryu
+-- algorithm
+-- NB: only valid if not NaN, +/-0, or +/-Inf. In practice, all calls should
+--     guarded by `specialStr`
+{-# INLINE ryu_d2s_fd #-}
+ryu_d2s_fd :: Double -> FloatingDecimal64
+ryu_d2s_fd f = unsafeDupablePerformIO $
+    alloca $ \mOut -> do
+      alloca $ \eOut -> do
+        c_ryu_d2s_fd (realToFrac f) mOut eOut
+        m <- peek mOut
+        e <- peek eOut
+        return $ FD64 m e
+
+asCBool :: Bool -> CUChar
+asCBool x = if x then 1 else 0
+
+{-# INLINE ryu_f2s_to_chars #-}
+ryu_f2s_to_chars :: Word32 -> Int32 -> Bool -> ByteString
+ryu_f2s_to_chars m e s = unsafeDupablePerformIO $ do
+    fp <- mallocByteString f2s_max_digits :: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_f2s_to_chars (fromIntegral m) (fromIntegral e) (asCBool s) p
+
+{-# INLINE ryu_d2s_to_chars #-}
+ryu_d2s_to_chars :: Word64 -> Int32 -> Bool -> ByteString
+ryu_d2s_to_chars m e s = unsafeDupablePerformIO $ do
+    fp <- mallocByteString d2s_max_digits :: IO (ForeignPtr Word8)
```

Comment:
Something like `boundedPrim 20` would make it even faster, because it doesn't allocate a ByteString. Same goes for `ryu_f2s`, `ryu_d2s` and `ryu_f2s_to_chars`.

Reply:
@Lumaere Can you make them return `BoundedPrim` instead of `ByteString`? I think it's going to make a big difference

Reply:
Sure. Added in https://github.com/haskell/bytestring/pull/222/commits/b3a033dfe38ebc6138bf22acf295f4dc02b1d769. The new results on my PC are
```
benchmarking Data.ByteString.Builder/Non-bounded encodings/foldMap floatDec (10000)
time                 992.7 μs   (984.3 μs .. 1.001 ms)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 987.9 μs   (982.7 μs .. 998.0 μs)
std dev              22.92 μs   (12.41 μs .. 36.71 μs)
variance introduced by outliers: 12% (moderately inflated)

benchmarking Data.ByteString.Builder/Non-bounded encodings/foldMap doubleDec (10000)
time                 1.071 ms   (1.066 ms .. 1.076 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.073 ms   (1.070 ms .. 1.076 ms)
std dev              11.26 μs   (8.056 μs .. 15.52 μs)
```

Reply:
Nice!